### DISCUSSION TITLE
Is it possibly to dynamic map hl7 xml?

### DISCUSSION THREAD

**Discussion by jtyme** [nextgenhealthcare/connect#5899] [Q&A]
Sorry, I am not familiar with older XML syntax and used JSON more.

Not even sure if this is possible in mirth.  Is it possible to store the HL7 XML in (preferably DB table) and then used to get the value from a message. JSON or just a map is acceptable.

Concept would be (JSON FOR EXAMPLE) 
```JS
var messageMappings =  [{"FirstName" : "['PID']['PID.4']['PID.4.1']"}, {"LastName" : "['PID']['PID.4']['PID.4.2']"}] ;

messageMappings.forEach((obj) => {
	Object.keys(obj).forEach((key) => {
		var temp = msg.obj[key].toString();
		channelMap.put(key.toString(), temp);
	})
});
```

I have also considered storing the last value so above would just be 'PID.4.1' and then try to build the rest in code.
```JS
var segment = null;
messageMappings.forEach((obj) => {
	Object.keys(obj).forEach((key) => {
	      var segArray = obj[key].split('.');
               segArray.forEach((seg) => {
                   segment += [ "'" + seg + "'"]
               });
var temp = msg.obj[key].toString();
		channelMap.put(key.toString(), temp);
	})
});
```

**Comment by jonbartels** [comment]
I think what you want to do is possible. Can you please edit your post and put the code into code formatting tags so it is easier to read?

**Reply by jtyme** [reply]
I tried to update it but it looks like one of the sections didn't take correctly. Sorry

**Reply by pacmano1** [reply]
I would likely do the explicit assigments to your JSON object, only because your meaning for a given segment, field, repeated field may or may not match the senders intent.     Also, it's extremely rare you care about all fields in some HL7 segment.  

Sure for things like basic demographics, it's fine.  But I'd expect that the break rather quickly with more complex message types.

**Reply by pacmano1** [reply]
@jtyme for code fomratting do this:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/793125d7-797d-42c0-b811-de7abd3fb72e)

Example:

```JS
<some mirth code you have already prettied>
```

**Reply by jtyme** [reply]
@pacmano1  thank you so much for the tip on code snippet!!

For the explicit assignments in json object you talking about this? {"FirstName" : "['PID']['PID.4']['PID.4.1']"}

I will test and see if I can correct the syntax....i dont think this will work as shown cause in my loop the syntax would be msg.['PID']['PID.4']

so I am sure I would need to use some other XML parsing like msg.PID.PID.4?

**Reply by pacmano1** [reply]
I mean I would use explicit assigments:

```JS
tmp.patient =
{
  p_sendingfacility : msg['MSH']['MSH.4']['MSH.4.1'].toString(),
  p_mrn : msg['PID']['PID.3']['PID.3.1'].toString(),
  p_alternatemrn : msg['PID']['PID.4']['PID.4.1'].toString(),
  p_familyname : msg['PID']['PID.5']['PID.5.1'].toString(),
  p_givenname : msg['PID']['PID.5']['PID.5.2'].toString(),
  p_middlename : msg['PID']['PID.5']['PID.5.3'].toString(),
  p_prefix : '',
  p_suffix : '',
  p_dob : msg['PID']['PID.7']['PID.7.1'].toString() != '' ? msg['PID']['PID.7']['PID.7.1'].toString() : null,
  p_gender : msg['PID']['PID.8']['PID.8.1'].toString(),
  p_raceid : msg['PID']['PID.10']['PID.10.1'].toString(),
  p_racetext : msg['PID']['PID.10']['PID.10.2'].toString(),
  p_street : msg['PID']['PID.11']['PID.11.1'].toString(),
  p_city : msg['PID']['PID.11']['PID.11.3'].toString(),
  p_province : msg['PID']['PID.11']['PID.11.4'].toString(),
  p_zip : msg['PID']['PID.11']['PID.11.5'].toString(),
  p_countrycode : msg['PID']['PID.11']['PID.11.6'].toString(),
  p_phonenumberhome : msg['PID']['PID.13'][0]['PID.13.1'].toString(),
  p_emailaddresshome : msg['PID']['PID.13'][0]['PID.13.4'].toString(),
  p_phonenumberbusiness : '',
  p_emailaddressbusiness : msg['PID']['PID.14']['PID.14.4'].toString(),
  p_primarylanguageid : msg['PID']['PID.15']['PID.15.1'].toString(),
  p_primarylanguagetext : msg['PID']['PID.15']['PID.15.2'].toString(),
  p_maritalstatus : msg['PID']['PID.16']['PID.16.1'].toString(),
  p_religion : msg['PID']['PID.17']['PID.17.1'].toString(),
  p_ssn : msg['PID']['PID.19']['PID.19.1'].toString(),
  p_dln : msg['PID']['PID.20']['PID.20.1'].toString(),
  p_ethnicgroupid : msg['PID']['PID.22']['PID.22.3'].toString(),
  p_ethnicgrouptext : msg['PID']['PID.22']['PID.22.1'].toString(),
  p_publicityindicator : msg['PD1']['PD1.11']['PD1.11.1'].toString()
}
```

**Reply by pacmano1** [reply]
In my case that JSON is passed to code template that dynamically passes each value to paramters of a stored procedure and makes the db call.

**Reply by pacmano1** [reply]
I of course understand why you want to do what you asked in your OP,  I just think it will "break" fast due to data problems between your expectations and the senders intent.   That mapping I posted might be custom per EHR or custom per client, it just depends.

**Reply by jtyme** [reply]
And what your showing is what we have in place on some processes. Just when you are supporting 8 mirth appliances and about 8k and growing of channels its getting a little harder to maintain so trying to be able to standardize the mappings for reusability across.  Long term idea is to store mappings in a DB table that is maintained.

**Reply by pacmano1** [reply]
I would just store the HL7 message itself then if needed eventually because you forgot to parse something.

Whenever I hear 8k channels, I get really concerned about how folks designed their stuff.   That always sounds like channels per client vs multi-tenant channels.

**Reply by jtyme** [reply]
and we do store the full message....but we also store meta data used for searching for the full messages. And right now, we use sections of a message like the MRN, patient name and other fields to use for business logic. And yes I agree that design is very important and we have a mix bag....something yes are very granular and are broken down via client and locations. Until I solve this problem its hard for us to go to a multi-tenant design which is why I am trying to come up with design that would store the mappings for a specific tenant as we know not two tenants are the same and then based on that mapping parse the message for the same business logic. Currently there are multiple design issues I am trying to work with and trying to "standardize" as much as I can so we can get to a better multi-tenant process....with over million transactions a day we will never get there due to volumn but we can break out and standardize

**Comment by tonygermano** [comment]
The general idea will work, but you can't do `msg["['PID']['PID.4']['PID.4.1']"]` with xml as you couldn't if `msg` were a regular javascript object. Also, you need to consider what your syntax will look like for repeating segments and fields and how you'll handle those.

Lodash does work as a mirth code template, and the [get](https://lodash.com/docs/4.17.15#get) operation does what you want. I'm pretty sure it works with xml objects, too. That would allow you to 
`_.get(msg, "['PID']['PID.4']['PID.4.1']").toString()`
You'll still need to be careful of specifying which iteration you want for repeating segments and fields, or the get() will return multiple results as an XMLList.
